internal/controllers/gateway/gateway_utils.go

```diff
@@ -342,23 +343,76 @@ func getListenerStatus(
 		}
 		// there's no filter for protocols that don't use Hostname, but this won't be populated from earlier for those
 		if _, ok := conflictedHostnames[listener.Port][hostname]; ok {
-			reason = string(gatewayv1alpha2.ListenerReasonHostnameConflict)
+			conflictReason = string(gatewayv1alpha2.ListenerReasonHostnameConflict)
 		}
 
 		if _, ok := conflictedPorts[listener.Port]; ok {
-			reason = string(gatewayv1alpha2.ListenerReasonProtocolConflict)
+			conflictReason = string(gatewayv1alpha2.ListenerReasonProtocolConflict)
 		}
 
-		if len(reason) > 0 {
-			newConditions := []metav1.Condition{}
+		// If the listener uses TLS, we need to ensure that the gateway is granted to reference
+		// all the secrets it references
+		if listener.TLS != nil {
+			resolvedRefReason = string(gatewayv1alpha2.ListenerReasonResolvedRefs)
+			for _, certRef := range listener.TLS.CertificateRefs {
+				// only secrets are supported as certificate references
+				if (certRef.Group != nil && (*certRef.Group != "core" && *certRef.Group != "")) ||
+					(certRef.Kind != nil && *certRef.Kind != "Secret") {
+					resolvedRefReason = string(gatewayv1alpha2.ListenerReasonInvalidCertificateRef)
+					break
+				}
+				// if the certificate is in the same namespace of the gateway, no ReferenceGrant is needed
+				if certRef.Namespace == nil {
+					continue
+				}
+				// get the result of the certificate reference. If the returned reason is not successful, the loop
+				// must be broken because a secret reference isn't granted
+				resolvedRefReason = getReferenceGrantConditionReason(gateway.Namespace, certRef, referenceGrants)
+				if resolvedRefReason != string(gatewayv1alpha2.ListenerReasonResolvedRefs) {
+					break
+				}
+			}
+		}
+
+		newConditions := []metav1.Condition{}
+
+		// if resolvedRefReason has any value, it means that the listener references a secret.
+		// A ListenerConditionResolvedRefs condition must be set to reflect in the gateway status
+		// the outcome of that reference (that means if the gateway is granted to access that secret)
+		if resolvedRefReason != "" {
+			conditionStatus := metav1.ConditionTrue
+			if resolvedRefReason != string(gatewayv1alpha2.ListenerReasonResolvedRefs) {
+				conditionStatus = metav1.ConditionFalse
+			}
+			newConditions = append(newConditions,
+				metav1.Condition{
+					Type:               string(gatewayv1alpha2.ListenerConditionResolvedRefs),
+					Status:             conditionStatus,
+					ObservedGeneration: gateway.Generation,
+					LastTransitionTime: metav1.Now(),
+					Reason:             resolvedRefReason,
+				},
+				metav1.Condition{
+					Type:               string(gatewayv1alpha2.ListenerConditionReady),
+					Status:             conditionStatus,
+					ObservedGeneration: gateway.Generation,
+					LastTransitionTime: metav1.Now(),
+					Reason:             string(gatewayv1alpha2.ListenerReasonReady),
+					Message:            "the listener is not ready and cannot route requests",
```

Comment:
Unless the `resolvedRefReason` comes back with something other than `ListenerReasonResolvedRefs` this condition could be `metav1.ConditionTrue`, in which case this message doesn't make sense right? Should we not have a permutation of this message for the `ConditionTrue` case?